### DISCUSSION TITLE
Custom runtime chunk loader (because urls are unknown during build time)

### DISCUSSION THREAD

**Discussion by stupkad** [webpack/webpack#16023] [Q&A]
## Feature request

Load chunks from locations, which are unavailable during build time.

When uploading JavaScript assets to Hubspot CMS, the files are automatically minified, post-processed and pushed to a CDN, for caching reason Hubspot delivers each asset **from a different location**. So I am looking for a solution to load chunks from a location, which is **not known during build time**.

Find more details at the end.

**What is the expected behavior?**

Possibility to inject a custom chunk loader which provided the correct filename without the need of hacking the created JavaScript files.

**What is motivation or use case for adding/changing the behavior?**

During upload to Hubspot, JavaScript files are post processed and served from different locations for caching reasons.

**How should this be implemented in your opinion?**

Unclear.

**Are you willing to work on this yourself?**
yes

## More details

when uploading JavaScript assets to Hubspot CMS, the files are automatically minified, post-processed and pushed to a cdn, for caching reason Hubspot delivers each asset from a different location. So I am looking for a solution to load chunks from a location, which is not known during build time.

**Locally**

 - js/widget.js
 - js/41.widget.js
 - js/393.widget.js

**Remote**

- http://f.hubspotusercontent-eu1.net/hub/xxx/hub_generated/template_assets/50239061735/1656576587321/index.js
- http://f.hubspotusercontent-eu1.net/hub/xxx/hub_generated/template_assets/50238026977/1656576587779/41.widget.js
- http://f.hubspotusercontent-eu1.net/hub/xxx/hub_generated/template_assets/unknown/location/393.widget.js

And here the challenge starts, as 41.widget.js or 393.widget.js cannot be loaded, as they are not within the same url as index.js. Unfortunately, the final urls are not available during build time, the are created during upload to Hubspot.

During the upload to hubspot, some placeholders within the JavaScript assets are replaced (Hubspot Template language HubL). So, if I manage to define the output.publicPath **for each chunk** as ``https://{{get_asset_url('./js/<chunk>.js')}}`` Hubspot could replace this by the correct url. 

I tried to manually patch the created index.js which does the trick, but seems a bit risky for me, unstable and hard to automate.

```
!function () {

    /* much more */

    o.u = function (e) {
        console.log("Load chunk");
        console.log(e);
        var result = "";
        if (e === 41) result = "{{get_asset_url('/theme/js/widgets/41.widgets.js')}}";
        if (e === 393) result = "{{get_asset_url('/theme/js/widgets/393.widgets.js')}}";
        console.log(result);
        return result;

    }, o.p = "https://", 

    /* much more */
}();

```

**Comment by vankop** [comment]
> for caching reason Hubspot delivers each asset from a different location

thats is useless if you using `[hash]` ( or other hash placeholders ) in output.. maybe there is setting for this.. 

> So, if I manage to define the output.publicPath for each chunk as https://{{get_asset_url('./js/<chunk>.js')}} Hubspot could replace this by the correct url.

you can post process assets replacing `chunk.js` to `{{get_asset_url()}}` ( I'm afraid of using this directly in output config )

___________

Also problem with such urls, as I remember, if you using ( would like to use ) error logging you will mislead with stacktrace, e.g. with your files you will get stacktrace:
```
41.widget.js
widget.js
```

no info about version ( hash ) of this assets.. ( you will need to log unique app version or something like this in this case )

**Comment by stupkad** [comment]
unfortunately we cannot change Hubspot behaviour, even if useless (or stupid). We have to deal with that.

I have simplified the names in my post... For now there are only three chunks, but the project will grow...

But I have no clue where/how to replace chunk.js with {{get_asset_url()}}, as the loading seems to be a slightly complex process within function o.u(...).

Thanks in advance,
Dietmar

**Reply by vankop** [reply]
yeah, this is complex task.. you should write a webpack plugin that will do https://github.com/webpack/webpack/blob/main/lib/RuntimePlugin.js#L253 and add your runtime module like here https://github.com/webpack/webpack/blob/main/lib/RuntimePlugin.js#L266 , but modify output to get something like (maybe add a runtime module with `getPath` function) :
```js
/******/ 	/* webpack/runtime/get javascript chunk filename */
/******/ 	(() => {
/******/ 		// This function allow to reference async chunks
/******/ 		__webpack_require__.u = (chunkId) => {
/******/ 			// return url for filenames based on template
/******/ 			return __getPath__(chunkId) + ".widget.js";
/******/ 		};
/******/ 	})();
```
